src/coreComponents/linearAlgebra/utilities/ComponentMask.hpp

```diff
@@ -0,0 +1,501 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 Total, S.A
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+/**
+ * @file ComponentMask.hpp
+ */
+
+#ifndef GEOSX_LINEARALGEBRA_UTILITIES_COMPONENTMASK_HPP_
+#define GEOSX_LINEARALGEBRA_UTILITIES_COMPONENTMASK_HPP_
+
+#include "common/GeosxMacros.hpp"
+#include "common/Logger.hpp"
+
+#include <cstdint>
+
+namespace geosx
+{
+
+namespace internal
+{
+
+/// Helps make static_assert's condition dependent on a non-type template parameter.
+template< typename T >
+constexpr bool always_false( T )
+{
+  return false;
+}
+
+template< int N >
+struct ComponentMaskTypeImpl
+{
+  static_assert( always_false( N ), "Unsupported template parameter N for ComponentMask. Use a value between 8 and 64." );
+};
+
+/**
+ * @brief Macro for declaring a specialization of ComponentMaskTypeImpl template.
+ */
+#define GEOSX_LAI_COMPONENTMASK_DECL( N ) \
+  template<> \
+  struct ComponentMaskTypeImpl< N > \
+  { \
+    using type = std::uint ## N ## _t; \
+  }
+
+GEOSX_LAI_COMPONENTMASK_DECL( 8 );
+GEOSX_LAI_COMPONENTMASK_DECL( 16 );
+GEOSX_LAI_COMPONENTMASK_DECL( 32 );
+GEOSX_LAI_COMPONENTMASK_DECL( 64 );
+
+#undef GEOSX_LAI_COMPONENTMASK_DECL
+
+constexpr int roundToNextPowerOfTwo( int v )
+{
+  v--;
+  v |= v >> 1;
+  v |= v >> 2;
+  v |= v >> 4;
+  v |= v >> 8;
+  v |= v >> 16;
+  v++;
+  return v;
```

Comment:
does `v` need to be `unsigned` for this to always work?

Reply:
Yes, will fix